theseus/labs/lie/functional/se3_impl.py

```diff
@@ -161,197 +155,139 @@ def identity(
         near_zero, 1.0 / 6 - theta2 / 120, (theta - sine) / theta3_nz
     )
-
-    ret[:, :, 3:] = sine_by_theta * tangent_vector_lin
-    ret[:, :, 3:] += one_minus_cosine_by_theta2 * torch.cross(
-        tangent_vector_ang, tangent_vector_lin, dim=1
+    ret[..., 3:] = sine_by_theta * tangent_vector_lin
+    ret[..., 3:] += one_minus_cosine_by_theta2 * torch.cross(
+        tangent_vector_ang, tangent_vector_lin, dim=-2
     )
-    ret[:, :, 3:] += theta_minus_sine_by_theta3_t * (
-        tangent_vector_ang @ (tangent_vector_ang.transpose(1, 2) @ tangent_vector_lin)
+    ret[..., 3:] += theta_minus_sine_by_theta3_t * (
+        tangent_vector_ang @ (tangent_vector_ang.transpose(-1, -2) @ tangent_vector_lin)
     )
 
-    # compute jacobians
-    theta3_nz = theta_nz * theta2_nz
-    theta_minus_sine_by_theta3_rot = torch.where(
-        near_zero, torch.zeros_like(theta), theta_minus_sine_by_theta3_t
-    )
-    jac = tangent_vector.new_zeros(
-        tangent_vector.shape[0],
-        6,
-        6,
+    return ret, (
+        theta,
+        theta2_nz,
+        sine_by_theta,
+        one_minus_cosine_by_theta2,
+        theta_minus_sine_by_theta3_t,
     )
-    jac[:, :3, :3] = (
-        theta_minus_sine_by_theta3_rot
-        * tangent_vector_ang.view(-1, 3, 1)
-        @ tangent_vector_ang.view(-1, 1, 3)
-    )
-    diag_jac = jac.diagonal(dim1=1, dim2=2)
-    diag_jac += sine_by_theta.view(-1, 1)
 
-    jac_temp_rot = one_minus_cosine_by_theta2.view(-1, 1) * tangent_vector_ang.view(
-        -1, 3
-    )
 
-    jac[:, 0, 1] += jac_temp_rot[:, 2]
-    jac[:, 1, 0] -= jac_temp_rot[:, 2]
-    jac[:, 0, 2] -= jac_temp_rot[:, 1]
-    jac[:, 2, 0] += jac_temp_rot[:, 1]
-    jac[:, 1, 2] += jac_temp_rot[:, 0]
-    jac[:, 2, 1] -= jac_temp_rot[:, 0]
+def _exp_impl(tangent_vector: torch.Tensor) -> torch.Tensor:
+    check_tangent_vector(tangent_vector)
+    ret, _ = _exp_impl_helper(tangent_vector)
+    return ret
 
-    jac[:, 3:, 3:] = jac[:, :3, :3]
 
-    minus_one_by_twelve = torch.tensor(
-        -1 / 12.0,
-        dtype=sine_by_theta.dtype,
-        device=sine_by_theta.device,
-    )
+def _jexp_impl_helper(
+    tangent_vector: torch.Tensor,
+    rotation: torch.Tensor,
+    theta: torch.Tensor,
+    theta2_nz: torch.Tensor,
+    sine_by_theta: torch.Tensor,
+    one_minus_cosine_by_theta2: torch.Tensor,
+    theta_minus_sine_by_theta3_t: torch.Tensor,
+    theta_minus_sine_by_theta3_rot: torch.Tensor,
+):
+    size = tangent_vector.shape[:-1]
+    jac = tangent_vector.new_zeros(*size, 6, 6)
+
+    # compute rotation jacobians
+    jac[..., :3, :3], _ = SO3._jexp_impl_helper(
+        tangent_vector[..., 3:],
+        sine_by_theta,
+        one_minus_cosine_by_theta2,
+        theta_minus_sine_by_theta3_rot,
+    )
+    jac[..., 3:, 3:] = jac[..., :3, :3]
+
+    # compute translation jacobians
+    near_zero = theta < constants._SO3_NEAR_ZERO_EPS[tangent_vector.dtype]
+    minus_one_by_twelve = -tangent_vector.new_ones(1) / 12.0
```

Comment:
I think we can use `new_tensor()` here. Also, since this is only used in one place, might as well not make it a variable and pass it directly `torch.where()`.

Reply:
`new_tensor()` is not compatible with `functorch`. `minus_one_by_twelve = -1/12.0` seems to work for `PyTorch 2.0`.

Reply:
`torch.where(near_zero, -1/12.0, variable)` seems to work for both `PyTorch 1.13.1` and `PyTorch 2.0`. Wondering if we should replace them which is simpler.